src/NonCumulative/Soundness/LogRel.agda

```diff
@@ -38,335 +38,359 @@ data _⊢_∶N®_∈Nat : Ctx → Exp → D → Set where
 -- management.  One can find unfolded definitions using conjunctions in the comments
 -- near the definition of the gluing model.
 
+
 -- Gluing model for Π
-record ΠRel i Δ IT OT (σ : Subst)
-             (iA : A ≈ B ∈ 𝕌 i)
-             (RI : Ctx → Typ → Set)
-             (RO : ∀ {a a′} → a ≈ a′ ∈ El i iA → Ctx → Typ → Set)
-             (Rs : Ctx → Exp → Typ → D → Set) : Set where
+
+-- for reference, in the cumulative version
+-- the paper rule is 
+-- exists S T, Γ ⊢ R = Π S T ×
+--             Γ ⊢ S ® A ×
+--             ∀ (σ : Γ' ≤ Γ) s a, Γ' ⊢ s : S ® a ∈ A ⇒ Γ' ⊢ T [ σ ] · s ® F · a
+
+-- (1) Δ ⊢ IT [ σ ] ® iA [ RI ≔ _ ⊢ _ ® iA ]
+-- (2) ∀ {s a} (irel : Δ ⊢ s ∶ IT [ σ ] ® a ∈El i iA) (a∈ : a ∈′ El i iA) → Δ ⊢ OT [ σ , s ] ® ΠRT.T≈T′ (RT a∈) [ RS ≔ _ ⊢ _ ∶ _ ® _ ∈El i iA , RO ≔ (a∈ : a ∈′ El i iA) → _ ⊢ _ ® ΠRT.T≈T′ (RT a∈) ]
+record ΠRel i j k Δ IT OT (σ : Subst)
+            (i≡maxjk : i ≡ max j k)
+            (univ : ∀ {l} → l < i → Ty)
+            (jA : A ≈ B ∈ PERDef.𝕌 j (λ l<j → univ (ΠI≤ i≡maxjk l<j)))
+            (RI : Ctx → Typ → Set)
+            (RO : ∀ {a a′} → a ≈ a′ ∈ PERDef.El j (λ l<k → univ (ΠI≤ i≡maxjk l<k)) jA → Ctx → Typ → Set)
+            (Rs : Ctx → Exp → Typ → D → Set) : Set where
+  field
+    IT-rel : RI Δ (IT [ σ ])
+    OT-rel : Rs Δ s (IT [ σ ]) a → (a∈ : a ∈′ PERDef.El j (λ l<k → univ (ΠI≤ i≡maxjk l<k)) jA) → RO a∈ Δ (OT [ σ , s ∶ IT [ σ ] ↙ j ])
+
+-- ∃ IT OT,
+-- (1) Γ ⊢ Π (IT ↙ i) (OT ↙ j) ∶[ 1 + k ] Se k   
+-- (2) (IT ↙ i) ∷ Γ ⊢ OT ∶[ 1 + i ] Se i × 
+-- (3) ∀ {Δ σ} → Δ ⊢w σ ∶ Γ → ΠRel
+record GluΠ i j k Γ T {A B}
+            (i≡maxjk : i ≡ max j k)
+            (univ : ∀ {l} → l < i → Ty)
+            (jA : A ≈ B ∈ PERDef.𝕌 j (λ l<j → univ (ΠI≤ i≡maxjk l<j)))
+            (RI : Ctx → Typ → Set)
+            (RO : ∀ {a a′} → a ≈ a′ ∈ PERDef.El j (λ l<k → univ (ΠI≤ i≡maxjk l<k)) jA → Ctx → Typ → Set)
+            (Rs : Ctx → Exp → Typ → D → Set) : Set where
+  field
+    IT   : Typ
+    OT   : Typ
+    -- need these two props or they are too difficult to invert
+    ⊢IT  : Γ ⊢ IT ∶[ 1 + j ] Se j
+    ⊢OT  : (IT ↙ j) ∷ Γ ⊢ OT ∶[ 1 + k ] Se k
+    T≈   : Γ ⊢ T ≈ Π (IT ↙ j) (OT ↙ k) ∶[ 1 + i ] Se i
+    krip : Δ ⊢w σ ∶ Γ → ΠRel i j k Δ IT OT σ i≡maxjk univ jA RI RO Rs
+
+
+-- Gluing model for universes
+
+-- ...
+record GluU j i Γ t T A 
+            (i≡1+j : i ≡ 1 + j)
+            (univ : ∀ {l} → l < i → Ty)
+            (R : A ∈′ PERDef.𝕌 j (λ {l} l<j → univ {l} (<-trans l<j (≤-reflexive (sym i≡1+j)))) → Set) : Set where
```

Comment:
I think  `(R : A ∈′ univ ?)` might just be more convenient, and allows you to apply substitution more easily.

Reply:
not really? because rc in the ⟦_,_,_⟧_⊢_®_  requires a specific shape PERDef.𝕌 j univ, as in the args of (A ≈ B ∈ PERDef.𝕌 j univ)

Reply:
ok then